src/util/system.cpp

```diff
@@ -366,7 +370,7 @@ bool ArgsManager::ParseParameters(int argc, const char* const argv[], std::strin
             return false;
         }
 
-        std::optional<util::SettingsValue> value = InterpretValue(keyinfo, val, *flags, error);
+        std::optional<util::SettingsValue> value = InterpretValue(keyinfo, val ? &*val : nullptr, *flags, error);
```

Comment:
Why use `nullptr` in replace of `nullopt_t`?

Maybe
```diff
--- a/src/util/system.cpp
+++ b/src/util/system.cpp
@@ -236,7 +236,7 @@ KeyInfo InterpretKey(std::string key)
  * @return parsed settings value if it is valid, otherwise nullopt accompanied
  * by a descriptive error string
  */
-static std::optional<util::SettingsValue> InterpretValue(const KeyInfo& key, const std::string* value,
+static std::optional<util::SettingsValue> InterpretValue(const KeyInfo& key, const std::optional<std::string> value,
                                                          unsigned int flags, std::string& error)
 {
     // Return negated settings as false values.
@@ -370,7 +370,7 @@ bool ArgsManager::ParseParameters(int argc, const char* const argv[], std::strin
             return false;
         }
 
-        std::optional<util::SettingsValue> value = InterpretValue(keyinfo, val ? &*val : nullptr, *flags, error);
+        std::optional<util::SettingsValue> value = InterpretValue(keyinfo, val, *flags, error);
         if (!value) return false;
 
         m_settings.command_line_options[keyinfo.name].push_back(*value);
@@ -891,7 +891,7 @@ bool ArgsManager::ReadConfigStream(std::istream& stream, const std::string& file
         KeyInfo key = InterpretKey(option.first);
         std::optional<unsigned int> flags = GetArgFlags('-' + key.name);
         if (flags) {
-            std::optional<util::SettingsValue> value = InterpretValue(key, &option.second, *flags, error);
+            std::optional<util::SettingsValue> value = InterpretValue(key, option.second, *flags, error);
             if (!value) {
                 return false;
             }
```
?

And the combined diff will smaller.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/24830#discussion_r849563238

> Why use `nullptr` in replace of `nullopt_t`?
> ...
> And the combined diff will smaller.

Wow, I think you know me pretty well, because it is hard for me to resist a 2-line smaller diff! If other reviewers have a preference, I can change this, but I do like the plain pointer approach better. I generally don't like making copies of things that can be pointed to, especially since `val` in this case is a user-supplied value that doesn't have a predetermined size. I also like functions that take plain `T*` or `T&` types as inputs, because it means can just as efficiently be passed values whether they are coming from `std::optional` containers, other containers like `std::map`, or directly from the heap or stack. It's harder to break code up into functions that can be composed if functions assume unnecessary things about their inputs. These are just mild preferences, though, so can change if anyone feels more strongly.

Reply:
I think the pointer approach is better, even if it makes the diff slightly bigger.

Kinda surprised std::optional lacks an `operator T*` to return a pointer to its value or nullptr...